### PR TITLE
Render text along lines

### PR DESCRIPTION
**Pull request by ahocevar**
This pull request adds a `placement` property to `ol.style.Text`. Its default value, and what the renderer previously did, is `'point'`. When set to `'line'`, text for linestrings and multilinestrings will be rendered along the path of the line. For polygons and multipolygons, text will be rendered along the outer ring.

For more fine-grained control of text rendering along lines, there are also new `maxAngle` and `exceedLength` properties on `ol.style.Text`.

I updated the street-labels example to use the new `placement: 'line'`, option, which makes the example 99% simpler of course.

sep-10-2017 21-20-24

### CODE REVIEW

**Comment by weskamm** [comment]
Very impressive, labeling in ol4 is getting professional now!
Integrated label collision detection would be the next big thing i guess :-)

**Comment by ahocevar** [comment]
This is now ready for review.

**Comment by marcjansen** [comment]
Nice work, @ahocevar. A couple of notes:

* On my machine a test fails in Firefox (Probably related to the next bullet point):
```
08:26 $ npm test -- --browsers Firefox

> openlayers@4.3.2 pretest /home/jansen/eclipse-ws/ol
> npm run lint

> openlayers@4.3.2 lint /home/jansen/eclipse-ws/ol
> eslint tasks test src examples transforms

> openlayers@4.3.2 test /home/jansen/eclipse-ws/ol
> npm run karma -- --single-run "--browsers" "Firefox"

> openlayers@4.3.2 karma /home/jansen/eclipse-ws/ol
> node tasks/test.js start test/karma.config.js "--single-run" "--browsers" "Firefox"

12 09 2017 08:28:16.026:INFO [karma]: Karma v1.7.0 server started at http://0.0.0.0:9876/
12 09 2017 08:28:16.028:INFO [launcher]: Launching browser Firefox with unlimited concurrency
12 09 2017 08:28:16.035:INFO [launcher]: Starting browser Firefox
12 09 2017 08:28:18.913:INFO [Firefox 55.0.0 (Ubuntu 0.0.0)]: Connected on socket J3euwF9VLEEFluL2AAAA with id 66372565
Firefox 55.0.0 (Ubuntu 0.0.0) ol.rendering.style.Text #render renders multiline text with alignment options FAILED
	Error: expected 5.6899999999999995 to be below 5 (/home/jansen/eclipse-ws/ol/node_modules/expect.js/index.js:96)
Firefox 55.0.0 (Ubuntu 0.0.0) ol.rendering.style.Text #render renders multiline text with positioning options FAILED
	Error: expected 5.6899999999999995 to be below 5 (/home/jansen/eclipse-ws/ol/node_modules/expect.js/index.js:96)
```

* The example doesn't use a sans-serif font for me (Increased fontsize so the difference is more clear):
image
adding `sans-serif` as fallback to the `font` config fixes this.
image

**Comment by ahocevar** [comment]
Thanks for the feedback @marcjansen. I have added a sans-serif fallback to the street-labels example. The two other failing tests you have should not be related to this pull request. They should fail for you on master too. Right?

**Comment by marcjansen** [comment]
> The two other failing tests you have should not be related to this pull request. They should fail for you on master too. Right?

Testing right now.

**Comment by marcjansen** [comment]
This looks good to me, please merge at will, @ahocevar.

File: externs/olx.js [hunk 1]
```diff
@@ -7743,7 +7775,7 @@ olx.style.TextOptions.prototype.textAlign;
 
 /**
  * Text base line. Possible values: 'bottom', 'top', 'middle', 'alphabetic',
- * 'hanging', 'ideographic'. Default is 'alphabetic'.
+ * 'hanging', 'ideographic'. Default is 'middle'.
```
[the remark this answers is not in the document]
Comment (ahocevar) [reply externs/olx.js]: No, it was always like this. Just the documentation was wrong.

[the remark this answers is not in the document]
Comment (marcjansen) [reply externs/olx.js]: :+1:

**Comment by tschaub** [comment]
Nice work!

**Comment by jc101** [comment]
This is an excellent feature but may I ask if the line text rendering should work fine when using VectorTiles? I am finding that when a line crosses a vector boundary, I get two labels instead of one (using a text align = center).

**Comment by ahocevar** [comment]
If your tiles are generated with lines clipped at the tile boundary, then that's expected. However, you will soon be able to get clean rendering across tile boundaries with `declutter: true` set on the layer. See #7328.

**Comment by jc101** [comment]
Thanks for your comments. I must have an error somewhere because my lines are not, or at least don't seem to be, clipped at the boundaries and yet I still see the two labels. I don't think they are clipped because the feature has the same geometry in zoom levels where it crosses a tile boundary and those when it is fully contained within one tile.

I have tried an example where I have one line extending unclipped across two vector tiles (with a very large buffer on the tile and also a large renderBuffer, although I don't know how important that is). If I remove the feature from one of the tiles, when the other tile is displayed on a map I see the line cropped at the tile boundary, which I guess is expected. The label for the line is then positioned relative to the cropped line rather than the full line.

I'll keep looking to see if I can figure out the error or come up with an example to demonstrate.

**Comment by tomchadwin** [comment]
Am I right that text with `placement: 'line'` is placed at the beginning of the linestring, and there is no way to place it eg half way along?

**Comment by ahocevar** [comment]
No, you can control positioning along the line with the textAlign property.

**Pull request merged**